### PR TITLE
Functions for Cardano Simple

### LINKED ISSUE #91: Functions for cardano-simple
- Move `evalScript` from `TestSingleScript` to cardano-simple
- Implement new functions in cs:
  - ```
    utxoForTransaction :: 
      Map P.ScriptHash (C.Versioned P.Script) -> 
      Network ->
      Tx era -> 
      Either ToCardanoError (Ledger.UTxO era)
      ```
  - ```
    txBalance :: 
      PParams era -> 
      Map P.ScriptHash (C.Versioned P.Script) ->
      Network ->
      Tx era ->
      Ledger.Value era
    ```
  - ```
    evaluateScriptsInTx ::
      PParams era ->
      Map P.ScriptHash (C.Versioned P.Script) ->
      Network ->
      Tx era ->
      Either (TranslationError (Crypto era)) Alonzo.ExUnits
    ```
In an appropriate new or existing cs module.

Modify `getUtxo`, `checkBalance` and `checkUnits` in `Plutus.Model.Mock` to use them.

While you're at it, rename `checkUnits` to make it clear it is actually executing the scripts and validating the transaction.

### PR DESCRIPTION
**Pull request by Geometer1729**
Closes #91

### CODE REVIEW

File: psm/src/Plutus/Model/Mock.hs [hunk 1]
```diff
@@ -510,12 +510,12 @@ sendTx tx = do
 -}
 sendSingleTx :: Tx -> Run (Either FailReason Stat)
 sendSingleTx preTx@(Tx extra _) =
-  runValidate $
-    liftEither (processMints preTx) >>= \tx -> do
-      genParams <- gets (mockConfigProtocol . mockConfig)
-      case genParams of
-        AlonzoParams params -> checkSingleTx @Alonzo.Era params extra tx
-        BabbageParams params -> checkSingleTx @Babbage.Era params extra tx
+  runValidate $ do
```
Comment (Geometer1729) [inline psm/src/Plutus/Model/Mock.hs:513]: Just caught this while reading the module and it bugged me so I pulled the `>>= \tx ->` into the do block

File: cardano-simple/src/Cardano/Simple/Cardano/Babbage.hs [hunk 2]
```diff
@@ -57,7 +57,7 @@ type Era = BabbageEra StandardCrypto
 
 instance IsCardanoTx Era where
   getTxBody = C.body
-  toCardanoTx = toBabbageTx
+  toCardanoTx n p e tx = toBabbageTx (Plutus.txScripts tx) n p e tx
```
Comment (AriFordsham) [inline cardano-simple/src/Cardano/Simple/Cardano/Babbage.hs:60]: Can we push this down into `toBabbageTx`?

File: cardano-simple/src/Cardano/Simple/Eval.hs [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+  Ledger.PParams era ->
+  Ledger.CostModel ->
+  Plutus.SerialisedScript ->
+  [Ledger.Data era] ->
+  Maybe Plutus.EvaluationError
+evalScript lang pparams cm script args =
+  either Just (const Nothing) . snd $
+    Plutus.evaluateScriptCounting
+      (toPlutusLang lang)
+      (Alonzo.transProtocolVersion $ getField @"_protocolVersion" pparams)
```
Comment (AriFordsham) [inline cardano-simple/src/Cardano/Simple/Eval.hs:68]: Can we replace uses of `getField` with overloaded record dot syntax?
```suggestion
      (Alonzo.transProtocolVersion pparams._protocolVersion)
```

File: psm/src/Plutus/Model/Mock.hs [hunk 4]
```diff
@@ -598,58 +583,26 @@ checkSingleTx params extra tx = do
...
 
-    checkBalance :: Ledger.UTxO era -> Core.Tx era -> Validate ()
-    checkBalance utxo txBody =
+    checkBalance :: Validate ()
+    checkBalance = do
+      utxos <- gets mockUtxos
+      network <- gets $ mockConfigNetworkId . mockConfig
+      balance <- case txBalance @era utxos params network tx extra of
+        Left err -> throwError $ FailToCardano err
+        Right bal -> pure bal
```
Comment (AriFordsham) [inline psm/src/Plutus/Model/Mock.hs:592]:
```suggestion
        case txBalance @era utxos params network tx extra of
          Left err -> throwError $ FailToCardano err
          Right bal -> when
            (bal /= mempty)
            (throwError $ NotBalancedTx $ fromCardanoValue bal)
```

**Review by AriFordsham — CHANGES_REQUESTED** [review]
Overall very nice! Nice to see `ModelMock` getting tidier!

File: cardano-simple/src/Cardano/Simple/Cardano/Class.hs [hunk 5]
```diff
@@ -21,7 +21,6 @@ import PlutusLedgerApi.V2 qualified as P
 
 class (C.Crypto era ~ StandardCrypto) => IsCardanoTx era where
   toCardanoTx ::
-    Map P.ScriptHash (C.Versioned P.Script) ->
```
[the remark this answers is not in the document]
Comment (Geometer1729) [reply cardano-simple/src/Cardano/Simple/Cardano/Class.hs:24]: Oh sorry for jumping the gun on that. I though I remembered you saying yes must've been a different thread and Id didn't look closely or something.

**Comment by Geometer1729** [comment]
> Overall very nice! Nice to see `ModelMock` getting tidier!
> 
> I'm assuming you're confident no functional changes have resulted. Have you run the test suite to see if it catches anything?

Thanks :) . I'm reasonably confident I haven't changed the behavior, and yes the test suite still passes.

**Review by AriFordsham — APPROVED** [review]

**Pull request merged**